contracts/Shadow.sol

```diff
@@ -57,8 +59,53 @@ contract Shadow is IERC20, IERC20Metadata {
     }
 
     function transferFrom(address from, address to, uint256 amount) public override returns (bool) {
-        IShadowFactory(ORIGIN).safeTransferFromByShadow(msg.sender, from, to, ID, amount);
+        address spender = msg.sender;
+        _spendAllowance(from, spender, amount);
+        IShadowFactory(ORIGIN).safeTransferFromByShadow(spender, from, to, ID, amount);
         emit Transfer(from, to, amount);
         return true;
     }
+
+    function increaseAllowance(address spender, uint256 addedValue) public virtual returns (bool) {
+        address owner = msg.sender;
+        _approve(owner, spender, allowance(owner, spender) + addedValue);
+        return true;
+    }
+
+    function decreaseAllowance(address spender, uint256 subtractedValue) public virtual returns (bool) {
```

Comment:
`_approve`, `_spendAllowance`, `increaseAllowance` and `decreaseAllowance` should just return when `1155.approvedForAll` is true